phylo/src/alignment/alignment_builder.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+
+use anyhow::bail;
+
+use crate::align;
+use crate::alignment::{Alignment, InternalMapping, Mapping, PairwiseAlignment, Sequences};
+use crate::tree::{NodeIdx, NodeIdx::Internal as Int, NodeIdx::Leaf, Tree};
+use crate::Result;
+
+pub struct AlignmentBuilder<'a> {
+    tree: &'a Tree,
+    seqs: Sequences,
+    node_map: InternalMapping,
+}
+
+impl<'a> AlignmentBuilder<'a> {
+    pub fn new(tree: &'a Tree, seqs: Sequences) -> AlignmentBuilder<'a> {
+        AlignmentBuilder {
+            tree,
+            seqs,
+            node_map: InternalMapping::new(),
+        }
+    }
+
+    pub fn msa(mut self, msa: InternalMapping) -> Self {
+        self.node_map = msa;
+        self
+    }
+
+    pub fn build(self) -> Result<Alignment> {
+        if self.node_map.is_empty() {
+            if self.seqs.aligned {
+                self.build_from_seqs()
+            } else {
+                self.build_from_unaligned()
+            }
+        } else {
+            self.build_from_map()
+        }
+    }
+
+    /// This assumes that the tree structure matches the alignment structure and that the sequences are aligned.
+    fn build_from_seqs(self) -> Result<Alignment> {
```

Comment:
jfyi: usually I would say that `from_seqs`, `from_unaligned` and `from_map` is enough. not need to use verbs. here it's a builder, so just check whatever is recommended in rust

Reply:
Based on `HasMap::with_capacity` below, I'd say that `from_xxx` is the way to go

Reply:
No, I think this is worse. ChatGPT suggests using action-oriented and context-specific names, so this is the current version:
```
    pub fn build(self) -> Result<Alignment> {
        if !self.node_map.is_empty() {
            self.compile_from_map()
        } else if self.seqs.aligned {
            self.reconstruct_from_aligned_seqs()
        } else {
            self.align_unaligned_seqs()
        }
    }
```

I prefer this.